trino/client.py

```diff
@@ -906,67 +913,138 @@ def _double_map_func(self):
 
-    def _get_time_with_timezome(self, value, time_size, pattern):
-        matches = re.match(r'^(.*)([\+\-])(\d{2}):(\d{2})$', value)
+    def _map_time_timezone(self, time_size, time_format):
+        return lambda val: self._get_time_with_timezone(val, time_size, time_format)
+
+    def _map_time_trim_millis_digits(self, datetime_default_size, millis_to_trim_div, time_format):
+        return lambda val: datetime.strptime(val[:datetime_default_size]
+                                             + str(round(int(val[datetime_default_size:]) / millis_to_trim_div)),
+                                             time_format).time()
+
+    def _map_time(self, time_size, time_format):
+        return lambda val: datetime.strptime(val[:time_size], time_format).time()
+
+    def _get_time_with_timezone(self, value, time_size, pattern):
+        matches = re.match(r'^(?P<time>.*)(?P<sign>[\+\-])(?P<hours>\d{2}):(?P<minutes>\d{2})$', value)
         assert matches is not None
         assert len(matches.groups()) == 4
-        if matches.group(2) == '-':
-            tz = -timedelta(hours=int(matches.group(3)), minutes=int(matches.group(4)))
+        if matches.group('sign') == '-':
+            tz = -timedelta(hours=int(matches.group('hours')), minutes=int(matches.group('minutes')))
         else:
-            tz = timedelta(hours=int(matches.group(3)), minutes=int(matches.group(4)))
-        return datetime.strptime(matches.group(1)[:time_size], pattern).time().replace(tzinfo=timezone(tz))
+            tz = timedelta(hours=int(matches.group('hours')), minutes=int(matches.group('minutes')))
+        return datetime.strptime(matches.group('time')[:time_size], pattern).time().replace(tzinfo=timezone(tz))
 
-    def _get_number_of_digits(self, column):
-        args = column['arguments']
-        if len(args) == 0:
-            return 3, 0
-        ms_size = column['arguments'][0]['value']
-        if ms_size == 0:
-            return -1, 0
-        ms_to_trim = ms_size - min(ms_size, 6)
-        return ms_size, ms_to_trim
+    def _get_time_with_timezone_round_ms(self, value, time_size, millis_digits_to_trim, time_format):
+        matches = re.match(r'^(?P<time>.*)(?P<sign>[\+\-])(?P<hours>\d{2}):(?P<minutes>\d{2})$', value)
+        assert matches is not None
+        assert len(matches.groups()) == 4
+        if matches.group('sign') == '-':
+            tz = -timedelta(hours=int(matches.group('hours')), minutes=int(matches.group('minutes')))
+        else:
+            tz = timedelta(hours=int(matches.group('hours')), minutes=int(matches.group('minutes')))
+        time_str = matches.group('time')[:time_size]
+        millis_str = str(round(int(matches.group('time')[time_size:]) / millis_digits_to_trim))
+        return datetime.strptime(time_str + millis_str, time_format).time().replace(tzinfo=timezone(tz))
+
+
+class TimestampRowMapperFactory(AbstractTemporalRowMapperFactory):
+    def create(self, column, col_type):
```

Comment:
Add types and return types in all methods

Reply:
Fixed